### PR TITLE
chore: add sphinx configs and documentation

### PR DESCRIPTION
**Pull request by lost-theory**
Adds a basic sphinx config, CI action for building & publishing to `gh-pages`, and fills in a few docstrings.

This is one of the prerequisites for https://github.com/ApeWorX/ape/issues/21.

### CODE REVIEW

**Comment by lost-theory** [comment]
How to build a PDF (in case we want this at some point, needs a Docker container to work which we aren't using):

```
        - name: Build PDF artifact
          run: "apt-get update -y && apt-get install -y latexmk texlive-latex-recommended texlive-latex-extra texlive-fonts-recommended && sphinx-build -M latexpdf . ./_build"
        - name: Upload PDF artifact
          uses: actions/upload-artifact@v1
          with:
              name: Documentation
              path: docs/_build/latex/pdfexample.pdf
```

**Review by fubuloubu — APPROVED** [review]

File: .github/workflows/docs.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+          uses: actions/upload-artifact@v1
+          with:
+              name: DocumentationHTML
+              path: docs/_build/
+
+        # TODO: branch-specific folders for branches other than `main`
+        - name: Commit and publish documentation changes to gh-pages branch
+          run: |
+              git clone https://github.com/ApeWorX/eip712 --branch gh-pages --single-branch gh-pages
+              cp -r docs/_build/* gh-pages/
```
Comment (fubuloubu) [inline .github/workflows/docs.yaml:37]: Should this push to a `latest/` folder? (Then add `cp -r latest/ stable/` and `cp -r stable/ {{ TAG_NAME }}` to "on release" trigger)

**Review by fubuloubu — APPROVED** [review]

File: .github/workflows/docs.yaml (hunk 1 above)
Comment (lost-theory) [reply .github/workflows/docs.yaml:37]: Yes I'll do that before merging this.

Comment (lost-theory) [reply .github/workflows/docs.yaml:37]: This is now handled by the docs build script (the `python release_docs.py` command a few lines above).

File: .github/workflows/docs.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        - name: Setup Python
+          uses: actions/setup-python@v2
+          with:
+              python-version: 3.8
+
+        - name: Install Dependencies
+          run: pip install .[doc]
+
+        - name: Build HTML artifact
+          run: "python release_docs.py"
```
Comment (fubuloubu) [inline .github/workflows/docs.yaml:25]: why is this in quotes?

File: .github/workflows/docs.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        - name: Upload HTML artifact
+          uses: actions/upload-artifact@v1
+          with:
+              name: DocumentationHTML
+              path: docs/_build/
+
+        # TODO: branch-specific folders for branches other than `main`
+        - name: Commit and publish documentation changes to gh-pages branch
+          run: |
+              git clone https://github.com/lost-theory/eip712 --branch gh-pages --single-branch gh-pages
```
Comment (fubuloubu) [inline .github/workflows/docs.yaml:36]: Should it be like this?
```suggestion
              git clone https://github.com/ApeWorX/eip712 --branch gh-pages --single-branch gh-pages
```

**Review by fubuloubu — APPROVED** [review]

**Comment by lost-theory** [comment]
Ready for another review. Here's what's new since last time:

1. Add `sphinx-multiversion` to automatically handle building the docs for every tagged release (this is also used to build `stable` and `latest` versions).
2. Customized the RTD theme's `layout.html` to show a version picker. This required adding some custom JS and CSS.
3. Added a build script in Python instead of relying on shell commands in the Github action. This also lets us use the same build script in development vs. the CI job.
4. We are now deleting extra copies of large static files (10+MB worth of web fonts) that would have been added to every version of the docs (10 versions = 100+MB).

Ready for another review @fubuloubu.

**Comment by lost-theory** [comment]
You can see what the version picker looks like in the top left of the sidebar here:

https://lost-theory.github.io/eip712/stable/modules/eip712.validation.html

It selects the currently active docs version, shows all available versions (I don't have any tagged releases on my fork, but I tested locally that they do show up), and redirects you to the same page on the other version when you make a selection.

**Comment by fubuloubu** [comment]
> 4. We are now deleting extra copies of large static files (10+MB worth of web fonts) that would have been added to every version of the docs (10 versions = 100+MB).

Oh nice!!

**Comment by fubuloubu** [comment]
Beautiful! This looks great to me!

File: .github/workflows/docs.yaml (hunk 2 above)
Comment (lost-theory) [reply .github/workflows/docs.yaml]: I copied it from the other sphinx CI action I was looking into. Removed the quotes.

File: .github/workflows/docs.yaml (hunk 3 above)
Comment (lost-theory) [reply .github/workflows/docs.yaml]: Yes good catch! Updated it.

**Comment by fubuloubu** [comment]
Should `release_docs.py` be renamed to `build_docs.py`? I don't see any release actions it's taking there.

Also, it would be great to have a docs build action that runs on PR

**Comment by lost-theory** [comment]
> Should `release_docs.py` be renamed to `build_docs.py`? I don't see any release actions it's taking there.

Yeah I think originally was just using this for the stable/latest/release logic, then I just made it a more general build script. I renamed it to `build_docs.py`. 
 
> Also, it would be great to have a docs build action that runs on PR

Good idea! I added the pull request open and synchronized (commits pushed to PR) events to the docs action. It ran successfully on this branch.

**Pull request merged**